File: sanmao-api/scripts/start-local-tunnel.mjs
#!/usr/bin/env node
import fs from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';
import { spawn, spawnSync } from 'node:child_process';
import http from 'node:http';

function requestOk(url, timeoutMs) {
  return new Promise((resolve) => {
    const req = http.request(url, { method: 'GET' }, (res) => {
      res.resume();
      resolve((res.statusCode || 500) < 400);
    });
    req.on('error', () => resolve(false));
    req.setTimeout(timeoutMs, () => {
      req.destroy();
      resolve(false);
    });
    req.end();
  });
}

function isPidRunning(pid) {
  if (!pid) {
    return false;
  }
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

async function pathExists(targetPath) {
  try {
    await fs.access(targetPath);
    return true;
  } catch {
    return false;
  }
}

async function readPid(pidPath) {
  if (!(await pathExists(pidPath))) {
    return 0;
  }
  return Number((await fs.readFile(pidPath, 'utf8')).trim()) || 0;
}

function runCapture(command, args) {
  return spawnSync(command, args, {
    encoding: 'utf8',
    windowsHide: true,
    stdio: ['ignore', 'pipe', 'pipe'],
  });
}

function listenerPids(port) {
  const result = runCapture('netstat', ['-ano', '-p', 'tcp']);
  if (result.error || (result.status ?? 1) !== 0) {
    return [];
  }
  const pids = new Set();
  for (const line of result.stdout.split(/\r?\n/)) {
    const parts = line.trim().split(/\s+/);
    if (parts.length < 5 || parts[0].toUpperCase() !== 'TCP') {
      continue;
    }
    const localAddress = parts[1];
    const state = parts[3].toUpperCase();
    const pid = Number(parts[4]);
    if (state !== 'LISTENING' || !pid) {
      continue;
    }
    if (localAddress.endsWith(`:${port}`)) {
      pids.add(pid);
    }
  }
  return [...pids];
}

function killPid(pid) {
  if (!pid) {
    return;
  }
  runCapture('taskkill', ['/PID', String(pid), '/F', '/T']);
}

async function main() {
  const listenHost = process.env.SANMAO_TUNNEL_HOSTNAME || '127.0.0.1';
  const listenPort = Number(process.env.SANMAO_TUNNEL_PORT || '13000');
  const remoteHost = process.env.SANMAO_TUNNEL_HOST || 'root@120.24.144.153';
  const remoteTarget = process.env.SANMAO_TUNNEL_TARGET || '127.0.0.1:3000';
  const pidPath = process.env.SANMAO_TUNNEL_PID || path.join(os.homedir(), '.ssh', 'sanmao-tunnel.pid');
  const healthUrl = process.env.SANMAO_TUNNEL_HEALTH_URL || `http://${listenHost}:${listenPort}/api/status`;

  const existingPid = await readPid(pidPath);
  if (existingPid && !isPidRunning(existingPid)) {
    await fs.rm(pidPath, { force: true });
  }

  const healthyPid = await readPid(pidPath);
  if (healthyPid && isPidRunning(healthyPid) && await requestOk(healthUrl, 3000)) {
    console.log(`[tunnel] already running: pid=${healthyPid}`);
    console.log(`[tunnel] health check ok: ${healthUrl}`);
    return;
  }

  for (const pid of listenerPids(listenPort)) {
    killPid(pid);
    console.log(`[tunnel] removed listener pid=${pid}`);
  }

  await fs.mkdir(path.dirname(pidPath), { recursive: true });

  const sshArgs = [
    '-o', 'ExitOnForwardFailure=yes',
    '-o', 'ServerAliveInterval=30',
    '-o', 'ServerAliveCountMax=3',
    '-N',
    '-L', `${listenHost}:${listenPort}:${remoteTarget}`,
    remoteHost,
  ];

  console.log(`[tunnel] opening ${listenHost}:${listenPort} -> ${remoteTarget} via ${remoteHost}`);
  const proc = spawn('ssh', sshArgs, {
    detached: true,
    stdio: 'ignore',
    windowsHide: true,
  });
  proc.unref();
  await fs.writeFile(pidPath, `${proc.pid}\n`, 'utf8');

  for (let attempt = 0; attempt < 5; attempt += 1) {
    if (await requestOk(healthUrl, 3000)) {
      console.log(`[tunnel] ready: pid=${proc.pid}`);
      console.log(`[tunnel] health check ok: ${healthUrl}`);
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }

  killPid(proc.pid);
  throw new Error(`[tunnel] failed health check: ${healthUrl}`);
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});

File: sanmao-api/scripts/start-local-tunnel.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import signal
import socket
import subprocess
import sys
import time
import urllib.request
from pathlib import Path


def request_ok(url: str, timeout: float) -> bool:
    try:
        with urllib.request.urlopen(url, timeout=timeout) as response:
            return response.status < 400
    except Exception:
        return False


def is_pid_running(pid: int) -> bool:
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


def read_pid(pid_path: Path) -> int:
    if not pid_path.exists():
        return 0
    try:
        return int(pid_path.read_text(encoding='utf-8').strip())
    except Exception:
        return 0


def listener_pids(port: int) -> list[int]:
    result = subprocess.run(
        ['netstat', '-ano', '-p', 'tcp'],
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return []
    pids: set[int] = set()
    for raw_line in result.stdout.splitlines():
        parts = raw_line.split()
        if len(parts) < 5 or parts[0].upper() != 'TCP':
            continue
        local_address = parts[1]
        state = parts[3].upper()
        try:
            pid = int(parts[4])
        except ValueError:
            continue
        if state != 'LISTENING' or pid <= 0:
            continue
        if local_address.endswith(f':{port}'):
            pids.add(pid)
    return list(pids)


def kill_pid(pid: int) -> None:
    if pid <= 0:
        return
    subprocess.run(['taskkill', '/PID', str(pid), '/F', '/T'], check=False, capture_output=True, text=True)


def main() -> int:
    listen_host = os.environ.get('SANMAO_TUNNEL_HOSTNAME', '127.0.0.1')
    listen_port = int(os.environ.get('SANMAO_TUNNEL_PORT', '13000'))
    remote_host = os.environ.get('SANMAO_TUNNEL_HOST', 'root@120.24.144.153')
    remote_target = os.environ.get('SANMAO_TUNNEL_TARGET', '127.0.0.1:3000')
    pid_path = Path(os.environ.get('SANMAO_TUNNEL_PID', str(Path.home() / '.ssh' / 'sanmao-tunnel.pid')))
    health_url = os.environ.get('SANMAO_TUNNEL_HEALTH_URL', f'http://{listen_host}:{listen_port}/api/status')

    pid_value = read_pid(pid_path)
    if pid_value and not is_pid_running(pid_value):
        pid_path.unlink(missing_ok=True)

    healthy_pid = read_pid(pid_path)
    if healthy_pid and is_pid_running(healthy_pid) and request_ok(health_url, 3):
        print(f'[tunnel] already running: pid={healthy_pid}')
        print(f'[tunnel] health check ok: {health_url}')
        return 0

    for listener_pid in listener_pids(listen_port):
        kill_pid(listener_pid)
        print(f'[tunnel] removed listener pid={listener_pid}')

    pid_path.parent.mkdir(parents=True, exist_ok=True)

    ssh_args = [
        'ssh',
        '-o', 'ExitOnForwardFailure=yes',
        '-o', 'ServerAliveInterval=30',
        '-o', 'ServerAliveCountMax=3',
        '-N',
        '-L', f'{listen_host}:{listen_port}:{remote_target}',
        remote_host,
    ]
    print(f'[tunnel] opening {listen_host}:{listen_port} -> {remote_target} via {remote_host}')
    creationflags = getattr(subprocess, 'DETACHED_PROCESS', 0) | getattr(subprocess, 'CREATE_NEW_PROCESS_GROUP', 0)
    proc = subprocess.Popen(ssh_args, stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, creationflags=creationflags)
    pid_path.write_text(f'{proc.pid}\n', encoding='utf-8')

    for _ in range(5):
        if request_ok(health_url, 3):
            print(f'[tunnel] ready: pid={proc.pid}')
            print(f'[tunnel] health check ok: {health_url}')
            return 0
        time.sleep(1)

    kill_pid(proc.pid)
    raise SystemExit(f'[tunnel] failed health check: {health_url}')


if __name__ == '__main__':
    raise SystemExit(main())
